File: benchmarks/successor/representation-v1/transformed/deterministic/csharp/gold/007-query-engine-refactor/OrderFlowEngine.cs
namespace OrderFlow;

public sealed class Customer
{
    public string? Id { get; init; }
    public string? Tier { get; init; }
}

public sealed class Order
{
    public string Id { get; init; } = "";
    public DateTimeOffset CreatedAt { get; init; }
    public string Status { get; init; } = "";
    public int? Priority { get; init; }
    public DateTimeOffset? DueAt { get; init; }
    public Customer? Customer { get; init; }
}

public sealed class Request
{
    public string Operation { get; init; } = "";
    public Order?[]? Orders { get; init; }
    public DateTimeOffset? AsOf { get; init; }
    public string? Id { get; init; }
    public string? ToStatus { get; init; }
}

public sealed record Response(string[] Ids);

public sealed record TransitionResponse(string Id, string Status);

public static class OrderFlowEngine
{
    public static object Handle(Request loc_zk64hdzcrs)
    {
        if (fun_qbwdskpmpf(loc_zk64hdzcrs.Operation, "transition"))
        {
            return fun_2eafckheko(loc_zk64hdzcrs);
        }

        var orders = fun_svzzyc72ua(loc_zk64hdzcrs.Orders);
        IEnumerable<Order> selected;

        if (fun_qbwdskpmpf(loc_zk64hdzcrs.Operation, "ready"))
        {
            selected = fun_rtvkjd7gi7(orders);
        }
        else if (fun_qbwdskpmpf(loc_zk64hdzcrs.Operation, "overdue"))
        {
            var asOf = loc_zk64hdzcrs.AsOf
                ?? throw new InvalidOperationException("asOf is required for overdue");
            selected = fun_jjxhcduzec(orders, asOf);
        }
        else if (fun_qbwdskpmpf(loc_zk64hdzcrs.Operation, "atRisk"))
        {
            var asOf = loc_zk64hdzcrs.AsOf
                ?? throw new InvalidOperationException("asOf is required for atRisk");
            selected = fun_6q4qho5po5(orders, asOf);
        }
        else if (fun_qbwdskpmpf(loc_zk64hdzcrs.Operation, "vipReady"))
        {
            selected = fun_rtvkjd7gi7(orders).Where(fun_7j64sig4qk);
        }
        else
        {
            throw new InvalidOperationException($"Unknown operation: {loc_zk64hdzcrs.Operation}");
        }

        return new Response(selected.Select(loc_dagq2dkqyn => loc_dagq2dkqyn.Id).ToArray());
    }

    private static TransitionResponse fun_2eafckheko(Request loc_zk64hdzcrs)
    {
        if (string.IsNullOrEmpty(loc_zk64hdzcrs.Id))
        {
            throw new InvalidOperationException("id is required for transition");
        }

        if (string.IsNullOrEmpty(loc_zk64hdzcrs.ToStatus))
        {
            throw new InvalidOperationException("toStatus is required for transition");
        }

        var matches = fun_svzzyc72ua(loc_zk64hdzcrs.Orders)
            .Where(loc_dagq2dkqyn =>
                string.Equals(loc_dagq2dkqyn.Id, loc_zk64hdzcrs.Id, StringComparison.Ordinal))
            .Take(2)
            .ToArray();

        if (matches.Length == 0)
        {
            throw new InvalidOperationException("order not found for transition");
        }

        if (matches.Length > 1)
        {
            throw new InvalidOperationException("duplicate order id for transition");
        }

        var target = fun_53ppru7gia(matches[0].Status, loc_zk64hdzcrs.ToStatus);
        return new TransitionResponse(matches[0].Id, target);
    }

    private static string fun_53ppru7gia(string? source, string target)
    {
        if (fun_qbwdskpmpf(source, "pending"))
        {
            if (fun_qbwdskpmpf(target, "processing"))
            {
                return "processing";
            }

            if (fun_qbwdskpmpf(target, "cancelled"))
            {
                return "cancelled";
            }
        }
        else if (fun_qbwdskpmpf(source, "processing"))
        {
            if (fun_qbwdskpmpf(target, "completed"))
            {
                return "completed";
            }

            if (fun_qbwdskpmpf(target, "cancelled"))
            {
                return "cancelled";
            }
        }

        throw new InvalidOperationException("invalid transition");
    }

    private static Order[] fun_svzzyc72ua(Order?[]? orders) =>
        orders?.OfType<Order>().ToArray() ?? [];

    private static IEnumerable<Order> fun_rtvkjd7gi7(IEnumerable<Order> orders) =>
        orders
            .Where(loc_dagq2dkqyn => fun_qbwdskpmpf(loc_dagq2dkqyn.Status, "pending"))
            .OrderByDescending(loc_dagq2dkqyn => loc_dagq2dkqyn.Priority ?? 0)
            .ThenBy(loc_dagq2dkqyn => loc_dagq2dkqyn.CreatedAt)
            .ThenBy(loc_dagq2dkqyn => loc_dagq2dkqyn.Id, StringComparer.Ordinal);

    private static IEnumerable<Order> fun_jjxhcduzec(
        IEnumerable<Order> orders,
        DateTimeOffset asOf) =>
        orders
            .Where(loc_dagq2dkqyn =>
                fun_7hdzz5dlwr(loc_dagq2dkqyn.Status)
                && loc_dagq2dkqyn.DueAt.HasValue
                && loc_dagq2dkqyn.DueAt.Value < asOf)
            .OrderBy(loc_dagq2dkqyn => loc_dagq2dkqyn.DueAt!.Value)
            .ThenByDescending(loc_dagq2dkqyn => loc_dagq2dkqyn.Priority ?? 0)
            .ThenBy(loc_dagq2dkqyn => loc_dagq2dkqyn.Id, StringComparer.Ordinal);

    private static IEnumerable<Order> fun_6q4qho5po5(
        IEnumerable<Order> orders,
        DateTimeOffset asOf)
    {
        var upperBound = asOf.AddHours(24);

        return orders
            .Where(loc_dagq2dkqyn =>
                fun_7hdzz5dlwr(loc_dagq2dkqyn.Status)
                && loc_dagq2dkqyn.DueAt.HasValue
                && loc_dagq2dkqyn.DueAt.Value >= asOf
                && loc_dagq2dkqyn.DueAt.Value < upperBound)
            .OrderBy(loc_dagq2dkqyn => loc_dagq2dkqyn.DueAt!.Value)
            .ThenByDescending(loc_dagq2dkqyn => loc_dagq2dkqyn.Priority ?? 0)
            .ThenBy(loc_dagq2dkqyn => loc_dagq2dkqyn.Id, StringComparer.Ordinal);
    }

    private static bool fun_7j64sig4qk(Order loc_dagq2dkqyn) =>
        loc_dagq2dkqyn.Customer is not null
        && (fun_qbwdskpmpf(loc_dagq2dkqyn.Customer.Tier, "gold")
            || fun_qbwdskpmpf(loc_dagq2dkqyn.Customer.Tier, "platinum"));

    private static bool fun_7hdzz5dlwr(string? status) =>
        fun_qbwdskpmpf(status, "pending") || fun_qbwdskpmpf(status, "processing");

    private static bool fun_qbwdskpmpf(string? actual, string expected) =>
        string.Equals(actual, expected, StringComparison.OrdinalIgnoreCase);
}

File: benchmarks/successor/representation-v1/transformed/deterministic/fsharp/gold/007-query-engine-refactor/OrderFlowEngine.fs
module OrderFlowEngine

open System

[<CLIMutable>]
type Customer =
    { id: string
      tier: string }

[<CLIMutable>]
type Order =
    { id: string
      createdAt: DateTimeOffset
      status: string
      priority: Nullable<int>
      dueAt: Nullable<DateTimeOffset>
      customer: Customer }

[<CLIMutable>]
type Request =
    { operation: string
      orders: Order array
      asOf: Nullable<DateTimeOffset>
      id: string
      toStatus: string }

[<CLIMutable>]
type Response = { ids: string array }

[<CLIMutable>]
type TransitionResponse =
    { id: string
      status: string }

let private fun_qbwdskpmpf actual expected =
    String.Equals(actual, expected, StringComparison.OrdinalIgnoreCase)

let private fun_7hdzz5dlwr status =
    fun_qbwdskpmpf status "pending" || fun_qbwdskpmpf status "processing"

let private priorityOf (loc_dagq2dkqyn: Order) =
    if loc_dagq2dkqyn.priority.HasValue then loc_dagq2dkqyn.priority.Value else 0

let private fun_svzzyc72ua (orders: Order array) =
    if Object.ReferenceEquals(orders, null) then
        [||]
    else
        orders
        |> Array.choose (fun loc_dagq2dkqyn ->
            if isNull (box loc_dagq2dkqyn) then None else Some loc_dagq2dkqyn)

let private compareReady (left: Order) (right: Order) =
    let byPriority = compare (priorityOf right) (priorityOf left)

    if byPriority <> 0 then
        byPriority
    else
        let byCreatedAt = DateTimeOffset.Compare(left.createdAt, right.createdAt)

        if byCreatedAt <> 0 then
            byCreatedAt
        else
            StringComparer.Ordinal.Compare(left.id, right.id)

let private compareDuePriorityId (left: Order) (right: Order) =
    let byDueAt = DateTimeOffset.Compare(left.dueAt.Value, right.dueAt.Value)

    if byDueAt <> 0 then
        byDueAt
    else
        let byPriority = compare (priorityOf right) (priorityOf left)

        if byPriority <> 0 then
            byPriority
        else
            StringComparer.Ordinal.Compare(left.id, right.id)

let private fun_rtvkjd7gi7 (orders: Order array) =
    orders
    |> Array.filter (fun loc_dagq2dkqyn -> fun_qbwdskpmpf loc_dagq2dkqyn.status "pending")
    |> Array.sortWith compareReady

let private fun_jjxhcduzec asOf (orders: Order array) =
    orders
    |> Array.filter (fun loc_dagq2dkqyn ->
        fun_7hdzz5dlwr loc_dagq2dkqyn.status
        && loc_dagq2dkqyn.dueAt.HasValue
        && loc_dagq2dkqyn.dueAt.Value < asOf)
    |> Array.sortWith compareDuePriorityId

let private fun_6q4qho5po5 (asOf: DateTimeOffset) (orders: Order array) =
    let upperBound = asOf.AddHours(24.0)

    orders
    |> Array.filter (fun loc_dagq2dkqyn ->
        fun_7hdzz5dlwr loc_dagq2dkqyn.status
        && loc_dagq2dkqyn.dueAt.HasValue
        && loc_dagq2dkqyn.dueAt.Value >= asOf
        && loc_dagq2dkqyn.dueAt.Value < upperBound)
    |> Array.sortWith compareDuePriorityId

let private fun_7j64sig4qk (loc_dagq2dkqyn: Order) =
    not (isNull (box loc_dagq2dkqyn.customer))
    && (fun_qbwdskpmpf loc_dagq2dkqyn.customer.tier "gold"
        || fun_qbwdskpmpf loc_dagq2dkqyn.customer.tier "platinum")

let private fun_53ppru7gia source target =
    if fun_qbwdskpmpf source "pending" then
        if fun_qbwdskpmpf target "processing" then
            "processing"
        elif fun_qbwdskpmpf target "cancelled" then
            "cancelled"
        else
            raise (InvalidOperationException("invalid transition"))
    elif fun_qbwdskpmpf source "processing" then
        if fun_qbwdskpmpf target "completed" then
            "completed"
        elif fun_qbwdskpmpf target "cancelled" then
            "cancelled"
        else
            raise (InvalidOperationException("invalid transition"))
    else
        raise (InvalidOperationException("invalid transition"))

let private fun_2eafckheko (loc_zk64hdzcrs: Request) : TransitionResponse =
    if String.IsNullOrEmpty(loc_zk64hdzcrs.id) then
        raise (InvalidOperationException("id is required for transition"))

    if String.IsNullOrEmpty(loc_zk64hdzcrs.toStatus) then
        raise (InvalidOperationException("toStatus is required for transition"))

    let matches =
        loc_zk64hdzcrs.orders
        |> fun_svzzyc72ua
        |> Array.filter (fun loc_dagq2dkqyn ->
            String.Equals(loc_dagq2dkqyn.id, loc_zk64hdzcrs.id, StringComparison.Ordinal))

    if matches.Length = 0 then
        raise (InvalidOperationException("order not found for transition"))

    if matches.Length > 1 then
        raise (InvalidOperationException("duplicate order id for transition"))

    { id = matches[0].id
      status = fun_53ppru7gia matches[0].status loc_zk64hdzcrs.toStatus }

let private query (loc_zk64hdzcrs: Request) : Response =
    let orders = fun_svzzyc72ua loc_zk64hdzcrs.orders

    let selected =
        match loc_zk64hdzcrs.operation with
        | operation when fun_qbwdskpmpf operation "ready" ->
            fun_rtvkjd7gi7 orders
        | operation when fun_qbwdskpmpf operation "overdue" ->
            if not loc_zk64hdzcrs.asOf.HasValue then
                raise (InvalidOperationException("asOf is required for overdue"))

            fun_jjxhcduzec loc_zk64hdzcrs.asOf.Value orders
        | operation when fun_qbwdskpmpf operation "atRisk" ->
            if not loc_zk64hdzcrs.asOf.HasValue then
                raise (InvalidOperationException("asOf is required for atRisk"))

            fun_6q4qho5po5 loc_zk64hdzcrs.asOf.Value orders
        | operation when fun_qbwdskpmpf operation "vipReady" ->
            orders
            |> fun_rtvkjd7gi7
            |> Array.filter fun_7j64sig4qk
        | operation ->
            raise (InvalidOperationException($"Unknown operation: {operation}"))

    selected
    |> Array.map (fun loc_dagq2dkqyn -> loc_dagq2dkqyn.id)
    |> fun ids -> { ids = ids }

let handle (loc_zk64hdzcrs: Request) : obj | null =
    if fun_qbwdskpmpf loc_zk64hdzcrs.operation "transition" then
        box (fun_2eafckheko loc_zk64hdzcrs)
    else
        box (query loc_zk64hdzcrs)
